rfcs/2021-08-29-8381-vrl-iteration-support.md

```diff
@@ -0,0 +1,756 @@
+the actual key of the record within the object.
+
+This doesn't fit existing patterns in VRL. It looks as if there's a _dangling_
+variable `key` at the end that remains unused, but because we special-cased this
+situation, it would instead magically update the actual key in the object after
+the closure runs to completion.
+
+This can become more difficult to reason about if/when we introduce control-flow
+statements such as `break`, as you could have set `key` before calling `break`,
+which would then either still mutate the actual key, or not, depending on how we
+implement `break`, but either way, the program itself becomes less readable, and
+operators have to read the language documentation to understand the semantic
+differences between how code behaves _inside_ a function-closure and _outside_.
+
+Instead, the `map` function-closure is required to return a two-element array of
+`[key, value]`, which the function machinery then uses to update the actual
+values of the object record, e.g.:
+
+```coffee
+map(.) |key, value| {
+  key = upcase(key)
+
+  # The array return-value clearly defines the eventual key and value values.
+  [key, value]
+}
+```
+
+##### Tuple Type
+
+Alternatively, we could introduce a new `tuple` type to define the return-type
+of the closure:
+
+```coffee
+map(.) |key, value| {
+  key = upcase(key)
+
+  (key, value)
+}
+```
+
+They would semantically be the same, given that VRL supports mixing value types
+in arrays, which is usually the difference between a tuple and an array, that,
+and the fact that a tuple is immutable, but an array doesn't have to be.
+
+Semantically, using a tuple makes more sense, but it does add an extra `Value`
+type, and it does mean we have to convert that tuple to a type supported by
+JSON, which will likely be an array, so the external JSON representation of
+tuples remains the same, regardless of whether we use arrays or tuples inside
+VRL itself.
+
+Since there isn't a clear benefit at this moment to using a tuple over
+a two-element array, the choice is made to forgo adding the tuple type at this
+moment.
+
+#### Parser Ambiguity
+
+TODO, figure out how to handle parser ambiguity.
+
+That is, these are identical:
+
+```coffee
+# merge the result of `fn(a)` with `b` and `c`
+fn(a) | b | c
+```
+
+```coffee
+# call `fn(a)` with closure `|b| c`
+fn(a) |b| c
+```
+
+## Rationale
+
+Iteration unlocks solutions to many remapping scenarios we currently don't
+support. Not implementing this RFC would hold VRL back, and prevent operators
+with more complex use-cases from using Vector with VRL to achieve their goals.
+
+By adding iteration, we unlock the capability to resolve almost all use-cases in
+the future by introducing more iteration-based functions.
+
+## Drawbacks
+
+- It adds more complexity to the language.
+- There are potential performance foot guns when iterating over large
+  collections.
+- The parser and compiler have to become more complex to support this use-case.
```

Comment:
Just noting that all of these drawbacks are worth the cost to unblock the use cases iteration is solving.